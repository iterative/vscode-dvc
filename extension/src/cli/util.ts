import { EventEmitter } from 'vscode'
import { CliEvent, CliResult } from '.'
import { Process } from '../process/execution'

export const notifyStarted = (
  baseEvent: CliEvent,
  processStarted: EventEmitter<CliEvent>
): void => processStarted.fire(baseEvent)

export const notifyOutput = (
  process: Process,
  processOutput: EventEmitter<string>
): void => {
  process.all?.on('data', chunk =>
    processOutput.fire(
      (chunk as Buffer)
        .toString()
        .split(/(\r?\n)/g)
        .join('\r')
    )
  )
}

export const notifyCompleted = (
  { command, pid, cwd, duration, exitCode, stderr }: CliResult,
  processCompleted: EventEmitter<CliResult>
): void =>
  processCompleted.fire({
    command,
    cwd,
    duration,
    exitCode,
    pid,
    stderr: stderr?.replace(/\n+/g, '\n')
  })

export const captureStdErr = (process: Process): string => {
  let stderr = ''
  process.stderr?.on('data', chunk => (stderr += (chunk as Buffer).toString()))
  return stderr
}
