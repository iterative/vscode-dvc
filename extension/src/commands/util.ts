import { commands } from 'vscode'
import { RegisteredCommands } from './external'
import { Setup } from '../setup'
import { Context } from '../vscode/context'

export const showSetupOrExecuteCommand =
  <T>(setup: Setup, callback: (context: Context) => Promise<T | undefined>) =>
  (context: Context) =>
    setup.shouldBeShown()
      ? commands.executeCommand(RegisteredCommands.SETUP_SHOW_EXPERIMENTS)
      : callback(context)
