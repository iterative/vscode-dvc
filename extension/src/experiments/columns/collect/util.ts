import { Column, ColumnType } from '../../webview/contract'
import { Value } from '../../../cli/dvc/contract'
import { ConfigKey, getConfigValue } from '../../../vscode/config'

export type ColumnAccumulator = Record<string, Column>

const joinPathArray = (
  pathSegments: string[],
  sep: string,
  fileSep: string
): string[] => {
  const [fileSegment, ...rest] = pathSegments

  if (!fileSegment) {
    return []
  }

  if (rest.length === 0) {
    return [fileSegment]
  }

  return [fileSegment + fileSep + rest.join(sep)]
}

export const limitAncestorDepth = (
  ancestors: string[],
  sep: string,
  fileSep: string,
  label: string,
  limit = 5
): {
  limitedDepthAncestors: string[]
  limitedDepthLabel: string
} => {
  const [path, ...rest] = ancestors
  const collectedLimit = Number(
    getConfigValue(ConfigKey.EXP_TABLE_HEAD_MAX_LAYERS, limit)
  )

  switch (collectedLimit) {
    case 1:
      return {
        limitedDepthAncestors: [],
        limitedDepthLabel: joinPathArray([...ancestors, label], sep, fileSep)[0]
      }
    case 2:
      return {
        limitedDepthAncestors: joinPathArray(ancestors, sep, fileSep),
        limitedDepthLabel: label
      }
    default: {
      /* 
      The depth is only limited for the middle of the path array.
      The first and final layer are excluded, and the
      concatenated layer itself counts as one; because of this, we must subtract 3
      from what we want the final layer count to be.
      */
      const convertedLimit = collectedLimit - 3
      if (collectedLimit <= 0 || rest.length <= convertedLimit) {
        return { limitedDepthAncestors: ancestors, limitedDepthLabel: label }
      }
      const cutoff = rest.length - convertedLimit
      const concatenatedPath = rest.slice(0, cutoff).join(sep)
      return {
        limitedDepthAncestors: [path, concatenatedPath, ...rest.slice(cutoff)],
        limitedDepthLabel: label
      }
    }
  }
}

const mergeParentColumnByPath = (
  acc: ColumnAccumulator,
  type: ColumnType,
  path: string,
  parentPath: string,
  label: string
) => {
  if (!acc[path]) {
    acc[path] = {
      hasChildren: true,
      label,
      parentPath,
      path,
      type
    }
  } else {
    acc[path].hasChildren = true
  }
}

const getValueType = (value: Value) => {
  if (value === null) {
    return 'null'
  }
  if (Array.isArray(value)) {
    return 'array'
  }
  return typeof value
}

const setStringLength = (column: Column, stringLength: number) => {
  if (!column.maxStringLength || column.maxStringLength < stringLength) {
    column.maxStringLength = stringLength
  }
}

const mergeNumberIntoColumn = (column: Column, value: number) => {
  const { maxNumber, minNumber } = column
  if (maxNumber === undefined || maxNumber < value) {
    column.maxNumber = value
  }
  if (minNumber === undefined || minNumber > value) {
    column.minNumber = value
  }
}

export const mergeAncestors = (
  acc: ColumnAccumulator,
  type: ColumnType,
  path: string,
  limitedDepthAncestors: string[],
  join: (...pathArray: string[]) => string
) => {
  if (!acc[path]) {
    for (let i = 1; i <= limitedDepthAncestors.length; i++) {
      const pathArray = limitedDepthAncestors.slice(0, i)
      mergeParentColumnByPath(
        acc,
        type,
        join(...pathArray),
        join(...pathArray.slice(0, -1)),
        pathArray[pathArray.length - 1]
      )
    }
  }
}

const mergeValueIntoColumn = (column: Column, value: Value) => {
  const valueType = getValueType(value)
  if (!column.types) {
    column.types = [valueType]
  } else if (!column.types.includes(valueType)) {
    column.types.push(valueType)
  }

  setStringLength(column, String(value).length)

  if (valueType === 'number') {
    mergeNumberIntoColumn(column, value as number)
  }
}

const buildValueColumn = (
  path: string,
  parentPath: string,
  pathArray: string[],
  label: string,
  value: Value
) => {
  const valueType = getValueType(value)

  const newColumn: Column = {
    hasChildren: false,
    label,
    maxStringLength: String(value).length,
    parentPath,
    path,
    pathArray,
    type: pathArray[0] as ColumnType,
    types: [valueType]
  }

  if (valueType === 'number') {
    newColumn.maxNumber = value as number
    newColumn.minNumber = value as number
  }

  return newColumn
}

export const mergeValueColumn = (
  acc: ColumnAccumulator,
  path: string,
  parentPath: string,
  pathArray: string[],
  label: string,
  value: Value
) => {
  if (!acc[path]) {
    acc[path] = buildValueColumn(path, parentPath, pathArray, label, value)
    return
  }
  mergeValueIntoColumn(acc[path], value)
}
