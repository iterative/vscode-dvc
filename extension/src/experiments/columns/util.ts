import { ColumnType } from '../webview/contract'

export type SummaryAcc = {
  metrics: string[]
  params: string[]
}

export const collectFromColumnOrder = (path: string, acc: SummaryAcc) => {
  if (path.startsWith(ColumnType.METRICS)) {
    acc.metrics.push(path)
  } else if (path.startsWith(ColumnType.PARAMS)) {
    acc.params.push(path)
  }
}
