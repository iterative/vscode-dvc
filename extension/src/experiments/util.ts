import { CommitData } from './webview/contract'
import {
  EXPERIMENT_WORKSPACE_ID,
  ExpRange,
  ExpShowOutput,
  ExpState
} from '../cli/dvc/contract'
import { COMMITS_SEPARATOR } from '../cli/git/constants'

const formatCommitMessage = (commit: string) => {
  const lines = commit.split('\n').filter(Boolean)
  return `${lines[0]}${lines.length > 1 ? ' ...' : ''}`
}

const addDetailsToRevision = (
  acc: ExpShowOutput,
  out:
    | (ExpState & {
        experiments?: ExpRange[] | null | undefined
      })
    | undefined,
  branch: string,
  commit: CommitData | undefined
) => {
  if (!out) {
    return
  }
  acc.push({
    ...out,
    branch,
    commit,
    description: commit ? formatCommitMessage(commit.message) : undefined
  })
}

const collectCommitData = (
  acc: { [sha: string]: CommitData },
  commit: string
) => {
  const [sha, author, date, refNamesWithKey] = commit
    .split('\n')
    .filter(Boolean)

  if (!sha) {
    return
  }

  const commitData: CommitData = {
    author: author || '',
    date: date || '',
    message: (commit.match(/\nmessage:(.+)/s) || [])[1] || '',
    tags: []
  }

  if (refNamesWithKey) {
    const refNames = refNamesWithKey.slice('refNames:'.length)
    commitData.tags = refNames
      .split(', ')
      .filter(item => item.startsWith('tag: '))
      .map(item => item.slice('tag: '.length))
  }
  acc[sha] = commitData
}

const collectCommitsData = (output: string): { [sha: string]: CommitData } => {
  const acc: { [sha: string]: CommitData } = {}

  for (const commit of output.split(COMMITS_SEPARATOR)) {
    collectCommitData(acc, commit)
  }
  return acc
}

export const combineOutputs = (
  currentBranch: string,
  expShow: ExpShowOutput,
  gitLog: string,
  order: { branch: string; sha: string }[]
): ExpShowOutput => {
  const commits = collectCommitsData(gitLog)

  const outputByRev: {
    [rev: string]: ExpState & {
      experiments?: ExpRange[] | null | undefined
    }
  } = {}
  for (const out of expShow) {
    outputByRev[out.rev] = out
  }

  const acc: ExpShowOutput = [
    { ...outputByRev[EXPERIMENT_WORKSPACE_ID], branch: currentBranch }
  ]
  for (const { branch, sha } of order) {
    addDetailsToRevision(acc, outputByRev[sha], branch, commits[sha])
  }

  return acc
}
