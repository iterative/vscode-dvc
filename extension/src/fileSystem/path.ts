import { Uri } from 'vscode'

export const standardizePath = (path?: string): string | undefined => {
  if (!path) {
    return
  }
  return Uri.file(path).fsPath
}
