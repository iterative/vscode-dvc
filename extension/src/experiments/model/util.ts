import get from 'lodash.get'
import { formatDate } from '../../util/date'
import { splitColumnPath } from '../columns/paths'
import { Experiment } from '../webview/contract'

type Value = undefined | null | [] | string | number

const isDate = (value: Value): boolean =>
  !!(typeof value === 'string' && Date.parse(value))

const isLongNumber = (value: Value): boolean =>
  typeof value === 'number' && value.toString().length > 7

const getStringifiedValue = (value: Value): string => {
  if (Number.isNaN(value)) {
    return 'NaN'
  }

  if (isDate(value)) {
    return formatDate(value as string)
  }

  if (Array.isArray(value)) {
    return `[${value?.toString()}]`
  }

  if (value === undefined) {
    return '-'
  }

  if (isLongNumber(value)) {
    return (value as number).toPrecision(5)
  }

  return String(value)
}

export const getDataFromColumnPath = (
  experiment: Experiment,
  columnPath: string
): { splitUpPath: string[]; value: string | null } => {
  const splitUpPath = splitColumnPath(columnPath)
  const collectedVal = get(experiment, splitUpPath)
  const value = collectedVal?.value || collectedVal

  return {
    splitUpPath,
    value:
      columnPath === 'Created' && typeof value === 'undefined'
        ? null
        : getStringifiedValue(value)
  }
}
