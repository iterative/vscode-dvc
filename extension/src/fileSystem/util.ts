import { sep } from 'path'

export const getPathArray = (path: string): string[] => path.split(sep)

export const getPath = (pathArray: string[], idx: number) =>
  pathArray.slice(0, idx).join(sep)

export const getDirectChild = (pathArray: string[], idx: number) =>
  getPath(pathArray, idx + 1)

export const getParent = (pathArray: string[], idx: number) => {
  const parent = getPath(pathArray, idx - 1)
  if (!parent) {
    return undefined
  }
  return parent
}

export const removeTrailingSlash = (path: string): string =>
  path.endsWith(sep) ? path.slice(0, -1) : path
