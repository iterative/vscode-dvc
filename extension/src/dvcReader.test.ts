import { resolve, join } from 'path'
import { mocked } from 'ts-jest/utils'

import { inferDefaultOptions, getExperiments } from './dvcReader'
import fs from 'fs'
import { execPromise } from './util'
import complexExperimentsOutput from './webviews/experiments/complex-output-example.json'
import { PromiseWithChild } from 'child_process'

jest.mock('fs')
jest.mock('./util')

const mockedFs = mocked(fs)
const mockedExecPromise = mocked(execPromise)

const extensionDirectory = resolve(__dirname, '..')

const testReaderOptions = {
  bin: 'dvc',
  cwd: resolve()
}

beforeEach(() => {
  jest.resetAllMocks()
})

test('Inferring default options on a directory with accessible .env', async () => {
  mockedFs.accessSync.mockReturnValue()

  expect(await inferDefaultOptions(extensionDirectory)).toEqual({
    bin: join(extensionDirectory, '.env', 'bin', 'dvc'),
    cwd: extensionDirectory
  })
})

test('Inferring default options on a directory without .env', async () => {
  mockedFs.accessSync.mockImplementation(() => {
    throw new Error('Mocked access fail')
  })

  expect(await inferDefaultOptions(extensionDirectory)).toEqual({
    bin: 'dvc',
    cwd: extensionDirectory
  })
})

test('Command-mocked getExperiments matches a snapshot when parsed', async () => {
  mockedExecPromise.mockReturnValue(
    (Promise.resolve({
      stdout: JSON.stringify(complexExperimentsOutput),
      stderr: ''
    }) as any) as PromiseWithChild<{ stdout: string; stderr: string }>
  )

  expect(await getExperiments(testReaderOptions)).toMatchSnapshot()
})
