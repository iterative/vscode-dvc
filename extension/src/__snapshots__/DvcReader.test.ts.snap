// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command-mocked getExperiments matches a snapshot when parsed 1`] = `
Object {
  "42b8736b08170529903cd203a1f40382a4b4a8cd": Object {
    "baseline": Object {
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.4668000042438507,
          "loss": 1.9293040037155151,
          "val_accuracy": 0.5608000159263611,
          "val_loss": 1.8770883083343506,
        },
      },
      "name": "test-branch",
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:28:59",
    },
    "d3f4a0d3661c5977540d2205d819470cf0d2145a": Object {
      "checkpoint_parent": "f0778b3eb6a390d6f6731c735a2a4561d1792c3a",
      "checkpoint_tip": "d3f4a0d3661c5977540d2205d819470cf0d2145a",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.5672000050544739,
          "loss": 1.8168506622314453,
          "val_accuracy": 0.6463000178337097,
          "val_loss": 1.7643483877182007,
        },
      },
      "name": "exp-05694",
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2021-01-14T10:58:00",
    },
    "f0778b3eb6a390d6f6731c735a2a4561d1792c3a": Object {
      "checkpoint_parent": "f81f1b5a1248b9d9f595fb53136298c69f908e66",
      "checkpoint_tip": "d3f4a0d3661c5977540d2205d819470cf0d2145a",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.5672000050544739,
          "loss": 1.8168506622314453,
          "val_accuracy": 0.6463000178337097,
          "val_loss": 1.7643483877182007,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2021-01-14T10:57:59",
    },
    "f81f1b5a1248b9d9f595fb53136298c69f908e66": Object {
      "checkpoint_parent": "42b8736b08170529903cd203a1f40382a4b4a8cd",
      "checkpoint_tip": "d3f4a0d3661c5977540d2205d819470cf0d2145a",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.5202666521072388,
          "loss": 1.8722929954528809,
          "val_accuracy": 0.6096000075340271,
          "val_loss": 1.8196595907211304,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2021-01-14T10:57:53",
    },
  },
  "53c3851f46955fa3e2b8f6e1c52999acc8c9ea77": Object {
    "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d": Object {
      "checkpoint_parent": "22e40e1fa3c916ac567f69b85969e3066a91dda4",
      "checkpoint_tip": "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.5926499962806702,
          "loss": 1.775016188621521,
          "val_accuracy": 0.6704000234603882,
          "val_loss": 1.7233840227127075,
        },
      },
      "name": "exp-83425",
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:27:02",
    },
    "1ee5f2ecb0fa4d83cbf614386536344cf894dd53": Object {
      "checkpoint_parent": "53c3851f46955fa3e2b8f6e1c52999acc8c9ea77",
      "checkpoint_tip": "4fb124aebddb2adf1545030907687fa9a4c80e70",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.3723166584968567,
          "loss": 2.020392894744873,
          "val_accuracy": 0.4277999997138977,
          "val_loss": 1.9979370832443237,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.15,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2e-12,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": 3,
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:31:44",
    },
    "217312476f8854dda1865450b737eb6bc7a3ba1b": Object {
      "checkpoint_parent": "9523bde67538cf31230efaff2dbc47d38a944ab5",
      "checkpoint_tip": "42b8736b08170529903cd203a1f40382a4b4a8cd",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.4668000042438507,
          "loss": 1.9293040037155151,
          "val_accuracy": 0.5608000159263611,
          "val_loss": 1.8770883083343506,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:28:57",
    },
    "22e40e1fa3c916ac567f69b85969e3066a91dda4": Object {
      "checkpoint_parent": "91116c1eae4b79cb1f5ab0312dfd9b3e43608e15",
      "checkpoint_tip": "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.5926499962806702,
          "loss": 1.775016188621521,
          "val_accuracy": 0.6704000234603882,
          "val_loss": 1.7233840227127075,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:27:01",
    },
    "23250b33e3d6dd0e136262d1d26a2face031cb03": Object {
      "checkpoint_parent": "53c3851f46955fa3e2b8f6e1c52999acc8c9ea77",
      "checkpoint_tip": "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.40904998779296875,
          "loss": 1.9896177053451538,
          "val_accuracy": 0.49399998784065247,
          "val_loss": 1.9391471147537231,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:26:36",
    },
    "42b8736b08170529903cd203a1f40382a4b4a8cd": Object {
      "checkpoint_parent": "217312476f8854dda1865450b737eb6bc7a3ba1b",
      "checkpoint_tip": "42b8736b08170529903cd203a1f40382a4b4a8cd",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.4668000042438507,
          "loss": 1.9293040037155151,
          "val_accuracy": 0.5608000159263611,
          "val_loss": 1.8770883083343506,
        },
      },
      "name": "test-branch",
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:28:59",
    },
    "4fb124aebddb2adf1545030907687fa9a4c80e70": Object {
      "checkpoint_parent": "d1343a87c6ee4a2e82d19525964d2fb2cb6756c9",
      "checkpoint_tip": "4fb124aebddb2adf1545030907687fa9a4c80e70",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.3724166750907898,
          "loss": 2.0205044746398926,
          "val_accuracy": 0.4277999997138977,
          "val_loss": 1.9979370832443237,
        },
      },
      "name": "exp-e7a67",
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.15,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2e-12,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": 3,
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:31:52",
    },
    "90aea7f2482117a55dfcadcdb901aaa6610fbbc9": Object {
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": true,
      "timestamp": "2020-12-29T15:25:27",
    },
    "91116c1eae4b79cb1f5ab0312dfd9b3e43608e15": Object {
      "checkpoint_parent": "e821416bfafb4bc28b3e0a8ddb322505b0ad2361",
      "checkpoint_tip": "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.557449996471405,
          "loss": 1.8261293172836304,
          "val_accuracy": 0.6414999961853027,
          "val_loss": 1.7749212980270386,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:26:55",
    },
    "9523bde67538cf31230efaff2dbc47d38a944ab5": Object {
      "checkpoint_parent": "53c3851f46955fa3e2b8f6e1c52999acc8c9ea77",
      "checkpoint_tip": "42b8736b08170529903cd203a1f40382a4b4a8cd",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.4083833396434784,
          "loss": 1.9882521629333496,
          "val_accuracy": 0.4970000088214874,
          "val_loss": 1.9363881349563599,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:28:50",
    },
    "baseline": Object {
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.3484833240509033,
          "loss": 2.048856019973755,
          "val_accuracy": 0.4277999997138977,
          "val_loss": 1.9979369640350342,
        },
      },
      "name": "master",
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-11-21T19:58:22",
    },
    "c658f8b14ac819ac2a5ea0449da6c15dbe8eb880": Object {
      "checkpoint_parent": "23250b33e3d6dd0e136262d1d26a2face031cb03",
      "checkpoint_tip": "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.46094998717308044,
          "loss": 1.9329891204833984,
          "val_accuracy": 0.5550000071525574,
          "val_loss": 1.8825950622558594,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:26:43",
    },
    "d1343a87c6ee4a2e82d19525964d2fb2cb6756c9": Object {
      "checkpoint_parent": "1ee5f2ecb0fa4d83cbf614386536344cf894dd53",
      "checkpoint_tip": "4fb124aebddb2adf1545030907687fa9a4c80e70",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.3724166750907898,
          "loss": 2.0205044746398926,
          "val_accuracy": 0.4277999997138977,
          "val_loss": 1.9979370832443237,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.15,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2e-12,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": 3,
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:31:51",
    },
    "e821416bfafb4bc28b3e0a8ddb322505b0ad2361": Object {
      "checkpoint_parent": "c658f8b14ac819ac2a5ea0449da6c15dbe8eb880",
      "checkpoint_tip": "1ba7bcd6ce6154e72e18b155475663ecbbd1f49d",
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.5113166570663452,
          "loss": 1.8798457384109497,
          "val_accuracy": 0.6035000085830688,
          "val_loss": 1.827923059463501,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.124,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 5,
          "learning_rate": 2.1e-7,
          "log_file": "logs.csv",
          "process": Object {
            "threshold": 0.85,
          },
        },
      },
      "queued": false,
      "timestamp": "2020-12-29T15:26:49",
    },
  },
  "workspace": Object {
    "baseline": Object {
      "metrics": Object {
        "summary.json": Object {
          "accuracy": 0.4668000042438507,
          "loss": 1.9293040037155151,
          "val_accuracy": 0.5608000159263611,
          "val_loss": 1.8770883083343506,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "dropout": 0.122,
          "dvc_logs_dir": "dvc_logs",
          "epochs": 2,
          "learning_rate": 2.2e-7,
          "log_file": "logs.csv",
          "process": Object {
            "test_arg": "string",
            "threshold": 0.86,
          },
        },
      },
      "queued": false,
      "timestamp": null,
    },
  },
}
`;

exports[`Command-mocked runExperiment matches a snapshot 1`] = `"This is test DVC log output"`;
